gtsfm/view_graph_estimator/cycle_consistent_rotation_estimator.py

```diff
@@ -0,0 +1,237 @@
+"""A ViewGraphEstimator implementation which ensures relative rotations are consistent in the cycles of the graph.
+
+Authors: John Lambert, Ayush Baid
+"""
+from collections import defaultdict
+from enum import Enum
+from typing import Dict, List, Optional, Tuple
+
+import numpy as np
+from gtsam import Cal3Bundler, Rot3, Unit3, PinholeCameraCal3Bundler
+
+import gtsfm.utils.geometry_comparisons as comp_utils
+import gtsfm.utils.graph as graph_utils
+import gtsfm.utils.logger as logger_utils
+import gtsfm.utils.metrics as metrics_utils
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.common.view_graph import ViewGraph
+from gtsfm.evaluation.metrics import GtsfmMetric, GtsfmMetricsGroup
+from gtsfm.view_graph_estimator.view_graph_estimator_base import ViewGraphEstimatorBase
+
+logger = logger_utils.get_logger()
+
+ERROR_THRESHOLD = 7.0
+
+MAX_INLIER_MEASUREMENT_ERROR_DEG = 5.0
+
+
+class EdgeErrorAggregationCriterion(str, Enum):
+    """Aggregate cycle errors over each edge by choosing one of the following summary statistics:
+
+    MIN: Choose the mininum cycle error of all cyles this edge appears in. An edge that appears in ANY cycle
+        with low error is accepted. High recall, but can have low precision, as false positives can enter
+        (error was randomly cancelled out by another error, meaning accepted).
+    MEDIAN: Choose the median cycle error. robust summary statistic. At least half of the time, this edge
+        appears in a good cycle (i.e. of low error). Note: preferred over mean, which is not robust to outliers.
+
+    Note: all summary statistics will be compared with an allowed upper bound/threshold. If they exceed the
+    upper bound, they will be rejected.
+    """
+
+    MIN_EDGE_ERROR = "MIN_EDGE_ERROR"
+    MEDIAN_EDGE_ERROR = "MEDIAN_EDGE_ERROR"
+
+
+class CycleConsistentRotationViewGraphEstimator(ViewGraphEstimatorBase):
+    """A ViewGraphEstimator that filters two-view edges with high rotation-cycle consistency error.
```

Comment:
"A ViewGraphEstimator that filters two-view edges within a triplet with ..."

Reply:
I think the current docstring is actually a bit more accurate. We aren't filtering two-view edges that occur within a single triplet, but rather that occur across potentially many triplets.